Fix layout for sub-plot saved in pdf format

chengzhuzhang opened the pull request:
## Description

## Description

When saving sub panel as pdf format, problems are reported about cropped axis, lables. This PR aims to fixes sets that have subplot saving problems.
- Closes #970 meridional and zonal mean 2d plots, closes #294 (for enso) and other sets.
- tropcial_subseasonal: fixing padding border 
- aerosol_aeronet: fixing padding border and plotting
- enso_diags: fixing parser to make command-line work and fixing padding border
- QBO: fixing fig position and padding border
## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules

If applicable:

- [ ] New and existing unit tests pass with my changes (locally and CI/CD build)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have noted that this is a breaking change for a major release (fix or feature that would cause existing functionality to not work as expected)

chengzhuzhang commented on e3sm_diags/plot/utils.py:
```diff
@@ -38,7 +38,7 @@
 # (left, bottom, right, top) in page coordinates
 BorderPadding = Tuple[float, float, float, float]
-DEFAULT_BORDER_PADDING: BorderPadding = (-0.06, -0.03, 0.13, 0.03)
+DEFAULT_BORDER_PADDING: BorderPadding = (-0.07, -0.03, 0.13, 0.03)
```
This line extend the border of all pdf plots on the left side.

chengzhuzhang commented:
@wlin7 I fixed the pdf issue in this PR. The zonal mean plot can be found here with other fixed plots: https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.zhang40/tests/pdf_size/

chengzhuzhang commented on e3sm_diags/plot/aerosol_aeronet_plot.py:
```diff
@@ -108,4 +150,4 @@ def plot(
     plt.loglog(ref_site_arr, test_site_arr, "kx", markersize=3.0, mfc="none")
 
-    _save_plot(fig, parameter, PANEL_CFG, BORDER_PADDING)
+    _save_plot_aerosol_aeronet(fig, parameter)
```
Use local _save_plot function to flexibly accommodate two plots with very different layouts.

chengzhuzhang commented on e3sm_diags/plot/enso_diags_plot.py:
```diff
@@ -138,7 +167,7 @@ def plot_scatter(
     plt.legend()
 
-    _save_plot(fig, parameter)
+    _save_plot_scatter(fig, parameter)
```
the scatter plots has very different layout as 3 panel maps, use custom _save_plot_scatter here instead.

tomvothecoder commented on e3sm_diags/plot/enso_diags_plot.py:
```diff
@@ -41,6 +41,35 @@
+
+def _save_plot_scatter(fig: plt.Figure, parameter: EnsoDiagsParameter):
+    """Save the scatter plot using a simplified approach for single panel plots."""
+    import os
+
+    from e3sm_diags.driver.utils.io import _get_output_dir
```
Can move imports to the top of `enso_diags_plot.py`.

tomvothecoder commented on e3sm_diags/plot/aerosol_aeronet_plot.py:
```diff
@@ -22,7 +21,50 @@
+
+def _save_plot_aerosol_aeronet(fig, parameter):
+    """Save aerosol_aeronet plots with different border padding for each panel."""
+    import os
+
+    from matplotlib.transforms import Bbox
+
+    from e3sm_diags.driver.utils.io import _get_output_dir
```
Can move imports to top of `aerosol_aeronet_plot.py`.

tomvothecoder commented on e3sm_diags/plot/aerosol_aeronet_plot.py:
```diff
@@ -22,7 +21,50 @@
+def _save_plot_aerosol_aeronet(fig, parameter):
+    """Save aerosol_aeronet plots with different border padding for each panel."""
+    import os
+
+    from matplotlib.transforms import Bbox
+
+    from e3sm_diags.driver.utils.io import _get_output_dir
+
+    # Save the main plot
+    for f in parameter.output_format:
+        f = f.lower().split(".")[-1]
+        fnm = os.path.join(
+            _get_output_dir(parameter),
+            parameter.output_file + "." + f,
+        )
+        plt.savefig(fnm)
+        logger.info(f"Plot saved in: {fnm}")
+
+    # Save individual subplots with different border padding
+    border_paddings = [BORDER_PADDING_COLORMAP, BORDER_PADDING_SCATTER]
+
+    for f in parameter.output_format_subplot:
+        fnm = os.path.join(
+            _get_output_dir(parameter),
+            parameter.output_file,
+        )
+        page = fig.get_size_inches()
+
+        for idx, (panel, border_padding) in enumerate(zip(PANEL_CFG, border_paddings)):
+            # Extent of subplot
+            subpage = np.array(panel).reshape(2, 2)
+            subpage[1, :] = subpage[0, :] + subpage[1, :]
+            subpage = subpage + np.array(border_padding).reshape(2, 2)
+            subpage_list = list(((subpage) * page).flatten())
+            extent = Bbox.from_extents(*subpage_list)
+
+            # Save subplot
+            fname = fnm + ".%i." % idx + f
+            plt.savefig(fname, bbox_inches=extent)
+            logger.info(f"Sub-plot saved in: {fname}")
```
Much of the code in `_save_plot_aerosol_aeronet()` is identical to `utils.plot._save_plot()` (below).
https://github.com/E3SM-Project/e3sm_diags/blob/436456117022b59550c1d8188490e36d12f78662/e3sm_diags/plot/utils.py#L60-L124

I think the only difference  `for` loop on line 56. I suggest we extend `_save_plot()` (or break it up into smaller functions) to support this difference in order to keep the code maintainable.

tomvothecoder commented on e3sm_diags/plot/enso_diags_plot.py:
```diff
@@ -41,6 +41,35 @@
+
+    from e3sm_diags.driver.utils.io import _get_output_dir
+
+    for f in parameter.output_format:
+        f = f.lower().split(".")[-1]
+        fnm = os.path.join(
+            _get_output_dir(parameter),
+            parameter.output_file + "." + f,
+        )
+        plt.savefig(fnm)
+        logger.info(f"Plot saved in: {fnm}")
```
This block of code is identical in `_save_plot()`. We can break that part of as its own function and re-use it here, instead of copying the same code.

tomvothecoder reviewed (commented):
Hey @chengzhuzhang, thanks for the work. I suggest to make the `_save_plot_...()` functions more maintainable by re-using and extending the general `_save_plot()` or breaking the `_save_plot()` function into smaller pieces (something like `_save_main_plot()` and `_save_sub_plots()`.

Let me know if you need help implementing this.

chengzhuzhang commented:
@tomvothecoder thanks for the review. The new commit should address most of your comments. 

 ENSO Diags Set 

  - Moved inline imports to top of file
  - Refactored _save_plot_scatter() to use shared _save_single_subplot() helper
  - Preserved full-figure saving behavior
  - New: fixed newly discovered cropping issue.

  Aerosol AERONET Set 

  - Moved inline imports to top, removed unused Bbox import
  - Refactored _save_plot_aerosol_aeronet() to use shared helper function

  Tropical Subseasonal Set 

  - Completely eliminated duplicate _save_plot() function
  - Now imports and uses general _save_plot() from utils with custom parameters

  Shared Utilities 

  - Added new _save_single_subplot() helper function to utils.py
  - Supports both full-figure and panel-based subplot saving

chengzhuzhang commented:
I'm attaching command lines to reproduce figures include in https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.zhang40/tests/pdf_size_1/
```
e3sm_diags zonal_mean_2d --no_viewer --reference_data_path '/lcrc/soft/climate/e3sm_diags_data/obs_for_e3sm_diags/climatology' --test_data_path '/lcrc/group/e3sm/ac.zhang40/example_v3/v3.LR.historical_0051/post/atm/180x360_aave/clim/30yr' --results_dir '/lcrc/group/e3sm/public_html/diagnostic_output/ac.zhang40/tests/pdf_size_1' --case_id 'ERA5' --run_type 'model_vs_obs' --sets 'zonal_mean_2d' --variables 'T' --seasons 'ANN' --regions 'global' --plevs '50.0' '100.0' '150.0' '200.0' '250.0' '300.0' '350.0' '400.0' '450.0' '500.0' '550.0' '600.0' '650.0' '700.0' '750.0' '800.0' '850.0' '900.0' '950.0' '1000.0' --multiprocessing --num_workers '8' --main_title 'T ANN global' --backend 'cartopy' --output_format 'png' --canvas_size_w '1212' --canvas_size_h '1628' --figsize '8.5' '11.0' --dpi '150' --arrows --contour_levels '180' '185' '190' '200' '210' '220' '230' '240' '250' '260' '270' '280' '290' '295' '300' --test_name 'v3.LR.historical_0051' --short_test_name 'v3.LR.historical_0201' --test_colormap 'cet_rainbow.rgb' --ref_name 'ERA5' --reference_name 'ERA5 Reanalysis' --reference_colormap 'cet_rainbow.rgb' --diff_title 'Model - Observations' --diff_colormap 'diverging_bwr.rgb' --diff_levels '-7' '-6' '-5' '-4' '-3' '-2' '-1' '1' '2' '3' '4' '5' '6' '7' --granulate 'variables' 'seasons' 'regions' --selectors 'sets' 'seasons' --save_netcdf --output_format_subplot 'pdf'


e3sm_diags aerosol_aeronet --no_viewer --reference_data_path '/lcrc/soft/climate/e3sm_diags_data/obs_for_e3sm_diags/climatology' --test_data_path '/lcrc/group/e3sm/ac.zhang40/example_v3/v3.LR.historical_0051/post/atm/180x360_aave/clim/30yr' --results_dir '/lcrc/group/e3sm/public_html/diagnostic_output/ac.zhang40/tests/pdf_size_1' --case_id 'AERONET' --run_type 'model_vs_obs' --sets 'aerosol_aeronet' --variables 'AODVIS' --seasons 'ANN' --regions 'global' --regrid_tool 'xesmf' --regrid_method 'conservative_normed' --multiprocessing --num_workers '8' --backend 'mpl' --output_format 'png' --output_format_subplot 'pdf' --canvas_size_w '1212' --canvas_size_h '1628' --figsize '8.5' '11.0' --dpi '150' --arrows --contour_levels '0.0' '0.05' '0.1' '0.2' '0.3' '0.4' '0.5' '0.6' '0.7' '0.8' '0.9' '1.0' '1.1' '1.2' --test_name 'v3.LR.historical_0051' --short_test_name 'v3.LR.historical_0101' --test_colormap 'cet_rainbow.rgb' --ref_name 'AERONET' --reference_colormap 'cet_rainbow.rgb' --diff_title 'Model - Observations' --diff_colormap 'diverging_bwr.rgb' --granulate 'variables' 'seasons' 'plevs' 'regions' --selectors 'sets' 'seasons' --output_format_subplot 'pdf'

e3sm_diags enso_diags --no_viewer --reference_data_path '/lcrc/soft/climate/e3sm_diags_data/obs_for_e3sm_diags/time-series' --test_data_path '/lcrc/group/e3sm/ac.zhang40/example_v3/v3.LR.historical_0051/post/atm/180x360_aave/ts/monthly/5yr' --results_dir '/lcrc/group/e3sm/public_html/diagnostic_output/ac.zhang40/tests/pdf_size_1'  --case_id 'TAUX-feedback' --ref_timeseries_input --test_timeseries_input --run_type 'model_vs_obs' --sets 'enso_diags' --variables 'TAUX' --seasons 'ANN' --regions 'global' --regrid_tool 'xesmf' --regrid_method 'conservative_normed' --multiprocessing --num_workers '8' --main_title 'TAUX anomaly (NINO4) vs. TS anomaly (NINO3)' --backend 'mpl' --output_format 'png' --output_format_subplot 'pdf' --canvas_size_w '1212' --canvas_size_h '1628' --figsize '8.5' '11.0' --dpi '150' --arrows --test_name 'v3.LR.historical_0101' --short_test_name 'v3.LR.historical_0101' --test_colormap 'cet_rainbow.rgb' --ref_name 'ERA-Interim' --reference_name 'ERA-Interim' --reference_colormap 'cet_rainbow.rgb' --diff_title 'Model - Observations' --diff_colormap 'diverging_bwr.rgb' --granulate 'variables' 'plevs' 'regions' --selectors 'sets' 'seasons' --test_start_yr '1985' --test_end_yr '2014' --ref_start_yr '1979' --ref_end_yr '1989' --output_format_subplot 'pdf' --plot_type 'scatter'

e3sm_diags enso_diags --no_viewer --reference_data_path '/lcrc/soft/climate/e3sm_diags_data/obs_for_e3sm_diags/time-series' --test_data_path '/lcrc/group/e3sm/ac.zhang40/example_v3/v3.LR.historical_0051/post/atm/180x360_aave/ts/monthly/5yr' --results_dir '/lcrc/group/e3sm/public_html/diagnostic_output/ac.zhang40/tests/pdf_size_1' --case_id 'PRECT-response' --ref_timeseries_input --test_timeseries_input --run_type 'model_vs_obs' --sets 'enso_diags' --variables 'PRECT' --seasons 'ANN' --regions '20S20N' --regrid_tool 'xesmf' --regrid_method 'conservative_normed' --multiprocessing --num_workers '8' --main_title 'Regression coefficient, PRECT anomaly to NINO34' --backend 'mpl' --output_format 'png' --output_format_subplot 'pdf' --canvas_size_w '1212' --canvas_size_h '1628' --figsize '8.5' '11.0' --dpi '150' --arrows --contour_levels '-5' '-3' '-2' '-1' '-0.5' '-0.2' '0.2' '0.5' '1' '2' '3' '5' --test_name 'v3.LR.historical_0101' --short_test_name 'v3.LR.historical_0101' --test_colormap 'diverging_bwr.rgb' --ref_name 'GPCP_v2.3' --reference_name 'GPCP_v2.3' --reference_colormap 'diverging_bwr.rgb' --diff_title 'Model - Observations' --diff_colormap 'BrBG' --diff_levels '-5' '-3' '-2' '-1' '-0.5' '-0.2' '0.2' '0.5' '1' '2' '3' '5' --granulate 'variables' 'plevs' 'regions' --selectors 'sets' 'seasons' --test_start_yr '1985' --test_end_yr '2014' --ref_start_yr '1979' --ref_end_yr '1989' --output_format_subplot 'pdf'


e3sm_diags qbo --no_viewer --reference_data_path '/lcrc/group/e3sm/diagnostics/observations/Atm/time-series/' --test_data_path '/lcrc/group/e3sm/ac.zhang40/example_v3/v3.LR.historical_0051/post/atm/180x360_aave/ts/monthly/5yr' --results_dir '/lcrc/group/e3sm/public_html/diagnostic_output/ac.zhang40/tests/pdf_size_1' --case_id 'QBO-ERA-Interim' --ref_timeseries_input --test_timeseries_input --run_type 'model_vs_obs' --sets 'qbo' --variables 'U' --seasons 'ANN' 'DJF' 'MAM' 'JJA' 'SON' --regions 'global' --regrid_tool 'xesmf' --regrid_method 'conservative_normed' --multiprocessing --num_workers '8' --main_title 'QBO index, amplitude, and power spectral density for U' --backend 'cartopy' --output_format 'png' --canvas_size_w '1212' --canvas_size_h '1628' --figsize '8.5' '11.0' --dpi '150' --arrows --test_name 'v3.LR.historical_0051' --short_test_name 'v3.LR.historical_0051' --test_colormap 'cet_rainbow.rgb' --ref_name 'ERA-Interim' --reference_name 'ERA-Interim' --reference_colormap 'cet_rainbow.rgb' --diff_title 'Model - Observations' --diff_colormap 'diverging_bwr.rgb' --granulate 'variables' 'plevs' 'regions' --selectors 'sets' 'seasons' --test_start_yr '1985' --test_end_yr '2014' --ref_start_yr '1985' --ref_end_yr '2014' --start_yr '1985' --end_yr '2014' --output_format_subplot 'pdf'
```

wlin7 commented:
> @wlin7 I fixed the pdf issue in this PR. The zonal mean plot can be found here with other fixed plots: https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.zhang40/tests/pdf_size/

Thank you @chengzhuzhang . The subplots are good with this update.

tomvothecoder commented:
> @tomvothecoder thanks for the review. The new commit should address most of your comments.

Thanks @chengzhuzhang, I will review. 

The integration test with `arm_diags` breaks with `python >=3.10` and `xcdat=0.9.0` ([here](https://github.com/E3SM-Project/e3sm_diags/actions/runs/15430665012/job/43432036852#step:11:579)). I will investigate further.

```python
ERROR    e3sm_diags.parameter.core_parameter:core_parameter.py:353 Error in e3sm_diags.driver.arm_diags_driver
Traceback (most recent call last):
  File "/__w/e3sm_diags/e3sm_diags/e3sm_diags/parameter/core_parameter.py", line 350, in _run_diag
    single_result = module.run_diag(self)
                    ^^^^^^^^^^^^^^^^^^^^^
  File "/__w/e3sm_diags/e3sm_diags/e3sm_diags/driver/arm_diags_driver.py", line 57, in run_diag
    return _run_diag_annual_cycle(parameter)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/__w/e3sm_diags/e3sm_diags/e3sm_diags/driver/arm_diags_driver.py", line 281, in _run_diag_annual_cycle
    ds_ref_climo = ds_ref.temporal.climatology(target_var, "month")
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/github/home/miniconda3/envs/e3sm_diags_ci/lib/python3.11/site-packages/xcdat/temporal.py", line 645, in climatology
    return self._averager(
           ^^^^^^^^^^^^^^^
  File "/github/home/miniconda3/envs/e3sm_diags_ci/lib/python3.11/site-packages/xcdat/temporal.py", line 913, in _averager
    dv_avg = self._group_average(ds, data_var, skipna=skipna)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/github/home/miniconda3/envs/e3sm_diags_ci/lib/python3.11/site-packages/xcdat/temporal.py", line 1552, in _group_average
    self._weights = self._get_weights(time_bounds)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/github/home/miniconda3/envs/e3sm_diags_ci/lib/python3.11/site-packages/xcdat/temporal.py", line 1625, in _get_weights
    bounds_dim = bounds.get_bounds_dim(time_bounds)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/github/home/miniconda3/envs/e3sm_diags_ci/lib/python3.11/site-packages/xcdat/bounds.py", line 1044, in get_bounds_dim
    raise ValueError(
ValueError: Invalid bounds dimension. Supported dimensions include: ['bnds', 'bnd', 'bounds', 'bound'].
```

chengzhuzhang commented:
> The integration test with `arm_diags` breaks with `python >=3.10` and `xcdat=0.9.0` ([here](https://github.com/E3SM-Project/e3sm_diags/actions/runs/15430665012/job/43432036852#step:11:579)). I will investigate further.

Yes, I'm also wondering about the failed integration test in python >=3.10. Thanks for helping investigate!

xylar commented:
`nv` is used for vertices bounding a polygon and is not a reasonable bounds dimension for `time` in my view.

xylar commented:
That seems like a misuse of `nv` to me but it looks like xCDAT should be prepared for it.

chengzhuzhang commented:
@tomvothecoder it looks like using nv as the time bounds dimension is CF compliant. And because of this I would think data from ARM archive may adopt the same convention most of the places. In this case, we may need xCDAT to support it. For short term, we can temporarily renaming in e3sm_diags, or we can pin last version of xcdat, which can be a set back though.

tomvothecoder commented:
> @tomvothecoder it looks like using nv as the time bounds dimension is CF compliant. And because of this I would think data from ARM archive may adopt the same convention most of the places. In this case, we may need xCDAT to support it. For short term, we can temporarily renaming in e3sm_diags, or we can pin last version of xcdat, which can be a set back though.

In this PR, I'll add the temporary workaround in `e3sm_diags` to rename `nv` to an xCDAT supported name. I'll also open an issue in xCDAT to add `nv` to the supported list of bound dim names.

tomvothecoder commented on e3sm_diags/driver/arm_diags_driver.py:
```diff
@@ -278,6 +278,11 @@ def _run_diag_annual_cycle(parameter: ARMDiagsParameter) -> ARMDiagsParameter:
+                    # NOTE: The bounds dimension can be "nv", which is not
+                    # currently recognized as a valid bounds dimension
+                    # by xcdat. We rename it to "bnds" to make it compatible.
+                    ds_ref = _rename_bounds_dim(ds_ref)
```
Separate fix for integration test, unrelated to the PR.

tomvothecoder commented:
> Once you approve, I will update the order of the functions then merge. Also, waiting on the build to complete (it is stalled but Python 3.12 passed).

GitHub Actions is running into issues: 

https://www.githubstatus.com/

 Update - Actions is experiencing degraded availability. We are continuing to investigate.
Jun 05, 2025 - 18:35 UTC
Update - Users of Actions will see delays in jobs starting or job failures. Users of Pages will see slow or failed deployments
Jun 05, 2025 - 18:30 UTC

chengzhuzhang commented:
@tomvothecoder When I tested again after merging the PR, the pdf plots all return as eampty pdfs. What happened I think is that, before merging the PR I might have tested an early commit before the refactor commit [0398c99](https://github.com/E3SM-Project/e3sm_diags/pull/984/commits/0398c99e9c7b71ed33ebe9e09b918db4c1802427)[0398c99](https://github.com/E3SM-Project/e3sm_diags/pull/984/commits/0398c99e9c7b71ed33ebe9e09b918db4c1802427). It is not obvious if there is an issue with the refactor, but I think it is okay to revert to a state before the refactor and merge. Please let me know what you think. thanks.
